Draft: add multiple shape-based Chinese input methods

moonbeamcelery opened the pull request:
Added all tables from [here](https://github.com/fcitx/fcitx5-table-extra/tree/master/tables). Supports changing nlpProviders in subtype editor screen, and added locale variants to the locale selection screen.

TODO: a lot of constants need to be placed in proper places, right now it's hard-coded in some functions. Pointers on where to put those are appreciated, or we can post this question in upstream.

![cangjie](https://user-images.githubusercontent.com/114041522/198220615-1cc36627-e157-4172-8175-b71e8526e4b6.jpg)
![wubi](https://user-images.githubusercontent.com/114041522/198220674-26e28476-7ce4-412c-ac52-892564f6b411.jpg)
![locales](https://user-images.githubusercontent.com/114041522/198220768-3e9933fc-2dba-4a36-b203-023aef2693fe.jpg)

moonbeamcelery committed:
Draft: editor screen exposes nlpProviders and shape-based Chinese input methods as variants

Signed-off-by: moonbeamcelery <moonbeamcelery@proton.me>

moonbeamcelery committed:
Fix defaults for zhengma preset

Signed-off-by: moonbeamcelery <moonbeamcelery@proton.me>

moonbeamcelery committed:
Add word tables for added input methods

Signed-off-by: moonbeamcelery <moonbeamcelery@proton.me>

moonbeamcelery committed:
Fix: bug in zhengma preset

moonbeamcelery committed:
Draft: support composing with special characters by delegating nlpProvider to decide composing range.

moonbeamcelery commented:
Update: support for irregular set of characters some input methods use, like zhengma-pinyin here using "@" in additional to [a-z].

![zhengmapinyin](https://user-images.githubusercontent.com/114041522/198515595-05dd7654-09d2-46a9-83ef-2a357a39437e.jpg)

moonbeamcelery committed:
Catch SQLite errors such as layout (locale variant) not found (e.g. using HanShapeBased with JIS)

moonbeamcelery committed:
fixup: remove TODO

moonbeamcelery committed:
fix: partly addresses 2101, allow searching for locale in English for phones lacking system locale IME

moonbeamcelery commented on utils/convert_fcitx5_sqlite.py:
```diff
@@ -0,0 +1,196 @@
+
+# Execute in this folder to convert to sqlite:
+# https://github.com/fcitx/fcitx5-table-extra/tree/master/tables
+# The tables are in public domain per their README.
```
... although the repo itself is GPLv3.

Waelwindows commented:
Not sure if pinyin/ other phonetic systems are in scope for this PR. perhaps it's better to add them in as another provider? Probably better to do that after merging in [#2054](https://github.com/florisboard/florisboard/pull/2054).

As for the large table variants, not sure if we should add them in cause most of the characters in those don't even haven fonts on Android. However, i'm pretty sure there are some obscure Hanzi that do have fonts and are only mapped in the large tables. IIRC some variants of 乍/苲

moonbeamcelery commented:
On second thought, you're right. Maybe we can include those once upstream has an extension system for languages.

In light of a future extension, should we only do one or two of these, or should we get rid of *-large and *-pinyin and keep others?

Waelwindows commented:
> In light of a future extension, should we only do one or two of these, or should we get rid of *-large and *-pinyin and keep others?

We should definitely get rid of `pinyin` tables. I am inclined to remove the `large` tables however they seem to have some more trad stuff which could be useful

moonbeamcelery commented:
I think per chat discussion, it makes sense to change this system into an import-based one. Then it becomes trivial to just release multiple versions, e.g. one bare-bones, one adding everything.

moonbeamcelery committed:
Adds support for importing "language packs" (sqlite3 db for HanShapeBased for now)

moonbeamcelery committed:
Changes language pack to zip files. Adds a basic language pack class for storing metadata of IMEs.

Signed-off-by: moonbeamcelery <moonbeamcelery@proton.me>

moonbeamcelery committed:
Implement language pack as a type of Flex extension, and draft its import and view UI

Signed-off-by: moonbeamcelery <moonbeamcelery@proton.me>

moonbeamcelery commented:
Implemented the language pack as a type of Flex extension. 

Issues that could be polished in the future:
- Right now for the sqlite3 file, this tries to read from the sqlite3 file from a specific pack named `org.florisboard.hanshapebasedextendedpack`, and uses it to override the default one from `org.florisboard.hanshapebasedbasicpack`. Although the language tags from all `LanguagePackExtension`s will be made available in the language selection during subtype edit.
- `extensionManager.languagePacks.observeForever` does not work in `HanShapeBasedLanguageProvider` for some reason, and a hacky workaround is used
- The extension files including the huge tables are still copied each time `HanShapeBasedLanguageProvider.create()` is called.
- Why is there no `wubi.txt` `cangjie.txt` but only their `-large` versions in fcitx5? Anyway luckily they are small enough to include.

For example, you can import [this extension](https://drive.proton.me/urls/V8QYCCN4DG#mh9DPGdUxcNP) to instead use the ibus zhengma which @Waelwindows converted that includes traditional characters, but as this overrides the sqlite3 file, others like cangjie will not have working suggestions.

![languagepack](https://user-images.githubusercontent.com/114041522/200212238-d2032ccc-8747-48df-ab55-ec06184ad012.jpg)

moonbeamcelery commented:
Implemented the language pack as a type of Flex extension. 

Issues that could be polished in the future:
- Right now for the sqlite3 file, this tries to read from the sqlite3 file from a specific pack named `org.florisboard.hanshapebasedextendedpack`, and uses it to override the default one from `org.florisboard.hanshapebasedbasicpack`. Although the language tags from all `LanguagePackExtension`s will be made available in the language selection during subtype edit.
- `extensionManager.languagePacks.observeForever` does not work in `HanShapeBasedLanguageProvider` for some reason, and a hacky workaround is used
- The extension files including the huge tables are still copied each time `HanShapeBasedLanguageProvider.create()` is called.
- Why is there no `wubi.txt` `cangjie.txt` but only their `-large` versions in fcitx5? Anyway luckily they are small enough to include.

For example, you can import [this extension](https://drive.proton.me/urls/TZJSZ0A7JC#9ibqcOYy9o1J) to instead use the ibus zhengma which @Waelwindows converted that includes traditional characters, but as this overrides the sqlite3 file, others like cangjie will not have working suggestions.

Edit: fixed typos because my brain shorted

![langhagepack](https://user-images.githubusercontent.com/114041522/200232523-155e0d46-955f-4b4f-9276-764952618ce5.jpg)

moonbeamcelery commented:
Also [the extension version](https://drive.proton.me/urls/QJR03A7QWR#bPKQyz2ZaqZv) that includes everything from fcitx5-table-extra (does not work with the `zh_TW_zhengma` one).

I would appreciate it if anyone can proofread the translations especially for `cantonese.txt`, `cantonhk.txt`, and `jyutpingtable.txt`, or provide insight on which method is the most popular in HK.

moonbeamcelery committed:
fix: input method name translation

Signed-off-by: moonbeamcelery <moonbeamcelery@proton.me>

moonbeamcelery commented:
On second thought, the fcitx5 page never cited anything that wubi is in public domain. Maybe we leave it out? Alternatively, we can play it safe and not include anything at all in the main app, and put everything in the extension (may need to add a LICENSE in the zip and allow displaying it somewhere).

moonbeamcelery committed:
Trim down to zhengma, quickclassic, and cangjie for the barebones Chinese shape-based pack. Polish extension user documentation.

moonbeamcelery committed:
Fix hack to allow multiple language pack extensions to co-exist.

Signed-off-by: moonbeamcelery <moonbeamcelery@proton.me>

moonbeamcelery commented:
Update, I got rid of wubi for now in case it is not in public domain yet.
Update the default language pack with a link to a stub page (LANGUAGEPACK.md) where it can describe how additional packs can be downloaded.
Also now allowing multiple language packs to be imported and used at the same time.

Waelwindows commented:
Good work so far! 

I haven't been able to work on this PR as much as I wanted to. However, i think I should be able to devote some time now.
I've tried your work and it looks great! Here are my thoughts so far,
 
For one we shouldn't couple it with `HanShapeBased` too much as I'm anticipating that `HanShapeBased` will be completely refactored once TRIEs are in FlorisBoard. Also consider that `LanguageExtension` may be used for other languages than chinese in the future (e.g. Japanese)?

Second, perhaps we should extend `LanguageExtensions` a bit, maybe allow subtypes (cangjie layout has 汉字 on it just like the keyboard layout) + presets? The current behavior is a bit surprising for users as they can't see their imported language in the presets list. I'll try to look into this, perhaps in a different branch + PR.

As is, I feel like this PR is ready to be merged as is. I'll merge it soon if you don't have any complaints that is.

moonbeamcelery commented:
Thanks for the feedback! 

> For one we shouldn't couple it with HanShapeBased too much

[I agree](https://github.com/Waelwindows/florisboard/pull/4/files#diff-b17c6e596c837622d8fb1c8ecdec01819561cd8281437dfc418f0a9d54659cecR50). I tried to make it an abstract class and have multiple subclasses extend it, but I don't know how to make `LanguageExtensions` be polymorphic. The extension manager uses its `.serializer` which I can't get to work with subclasses. I'm pretty new to Kotlin.

On another hand, I see Patrick just put multiple functions of keyboard extensions (e.g. different composers, different layouts, etc) [in the same class](https://github.com/florisboard/florisboard/blob/master/app/src/main/kotlin/dev/patrickgold/florisboard/ime/keyboard/KeyboardExtension.kt#L36-L41), and have each extension only include one of them, so I think the same issue is there too.

> maybe allow subtypes (cangjie layout has 汉字 on it just like the keyboard layout) + presets

Do you mean instead of qwerty, cangjie actually has character parts on it? 

I like this idea, but I think it requires some extra work. Adding to presets is not implemented but I think should be easy?

Importing keyboard subtypes is present in the code but not there in the UI, so I'm not sure how well this is tested. But also importing a mixture of two kinds of extensions needs to be supported, and checks written. One way to do this is to re-organize this to make `languagepackExtension` a part of keyboard extensions, parallel to [everything else](https://github.com/florisboard/florisboard/blob/master/app/src/main/kotlin/dev/patrickgold/florisboard/ime/keyboard/KeyboardExtension.kt#L36-L41).

> I'll merge it soon if you don't have any complaints that is.

Please do, and thanks! Although if someone can [fix this](https://github.com/Waelwindows/florisboard/pull/4/files#diff-5e43739632a04b8041104a585aef8a1eaec4931c69dc15e21d472b2eac0250cfR168-R171) it would be great.

Waelwindows commented:
> > maybe allow subtypes (cangjie layout has 汉字 on it just like the keyboard layout) + presets
> 
> Do you mean instead of qwerty, cangjie actually has character parts on it?
> 
Yes, look at how the layout looks like on PCs! 

it has both but since the focus of the cangjie layout is on the cangjie input _itself_ i feel like it's appropriate to include. Especially since there's no set mobile layout for cangjie AFAIK unlike the JIS layout which i kind of regret doing now :sweat_smile: ~if not for the experience i got hacking on florisboard~

![image](https://user-images.githubusercontent.com/5228243/209566215-b7994378-7abe-4674-801f-02fc835bc1de.png)

> One way to do this is to re-organize this to make `languagepackExtension` a part of keyboard extensions, parallel to [everything else](https://github.com/florisboard/florisboard/blob/master/app/src/main/kotlin/dev/patrickgold/florisboard/ime/keyboard/KeyboardExtension.kt#L36-L41).

That seems like the better way to do it! Having everything consolidated under `KeyBoardExtension` gives freedom to the extension developer to include whatever they like. which may include more stuff in the future, and should simplify implementation on our end

> > I'll merge it soon if you don't have any complaints that is.
> 
> Please do, and thanks! Although if someone can [fix this](https://github.com/Waelwindows/florisboard/pull/4/files#diff-5e43739632a04b8041104a585aef8a1eaec4931c69dc15e21d472b2eac0250cfR168-R171) it would be great.

Since the above change seems to be feasible in scope for this PR, I think I'll merge after doing that first

moonbeamcelery commented:
Frankly, I didn't think about it well when I did it this way, because I remember Patrick was planning to add a language pack extension to FlorisBoard. I'd ask Patrick first on how he thinks about adding to KeyboardExtension and use it as the language pack extension, to make sure someone doesn't need to refactor it again in the near future. You can just ping on the FlorisBoard chat if you'd like.

Waelwindows commented:
> Update: support for irregular set of characters some input methods use, like zhengma-pinyin here using "@" in additional to [a-z].
> 
> ![zhengmapinyin](https://user-images.githubusercontent.com/114041522/198515595-05dd7654-09d2-46a9-83ef-2a357a39437e.jpg)

I haven't tried importing language packs before, but the `@` in `zhengma-pinyin` doesn't seem to to be included in the compositing range. I'm guessing that's why the pinyin suggestions don't pop up

![Screenshot_20221230-164533_FlorisBoard Debug](https://user-images.githubusercontent.com/5228243/210076818-55480563-eb34-4cf0-be6f-4be6beb57b05.png)

Just in case it was my old setup interfereing, i cleared all storage data for FlorisBoard debug and it didn't fix the issue. Any ideas?

moonbeamcelery commented:
Hmmm, it *looks* like that you typed a full-width `＠` instead of a half-width `@`? When you don't turn on Settings > Keyboard > Include symbols popup in letter key popup, then you only get the full-width `＠` in "a" key long-press. When you turn on that option, the two `＠@` can also be easily confused.

Some fix ideas:
- Just convert `＠` to `@` before querying.
- In fcitx, it seems you're able to set which character is used to indicate pinyin input. We don't have a language pack settings yet, though.
- How about we change the full- and half-width toggle into three states? All full, all half, and auto (full for punctuation, half for others). Popup uses auto.
    - Currently the cjk keyboard is different from other input methods I have used. The main popup for `@#%&-+*` are all full-width, and `" '` are half-width, which are confusing.
    - In desktop (Chinese) input methods, there are toggles for full- and half-width characters, and another full- and half-width toggle only for punctuation `",.;:?!'`.
    - Gboard doesn't have a full- and half-width toggle or symbol popups for letter keys, and the Chinese input defaults to full-width for `,.":;!?()'[]` with their half-width options for popups, and half-width for `@#$-+/\=~|*` and backtick. Only `-/~` have their full-width options in popups. Full-width for `@#$-+\=~|` and backtick cannot be typed. I tried its Japanese keyboards, and they largely follow this design. The Korean keyboard is mostly half-width only, and you cannot type full-width commas and periods.
    -
